File: knock out whist simulation/knock_out_whist.py
import random

class Card:
    def __init__(self, rank, suit):
        self.rank = rank
        self.suit = suit

    def __repr__(self):
        return f'{self.rank} of {self.suit}'

suits = ['Hearts', 'Diamonds', 'Clubs', 'Spades']
ranks = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A']

deck = [Card(rank, suit) for suit in suits for rank in ranks]

number_of_runs = 5000000
number_of_wins_for_trump = 0
for _ in range(number_of_runs):
    players = random.sample(deck, 4)
    #print(players)

    for i in range(1,4):
        #players[0] is trump
        if players[0].suit == players[i].suit:
            if players[0].rank < players[i].rank:
                #print("trump lost")
                break
    else:
        # triggers when loop ends without break
        number_of_wins_for_trump += 1
        #print("trump won")

print(f"Probability of winning: {number_of_wins_for_trump / number_of_runs}")

File: knock out whist simulation/knock_out_whist.hs
import System.Random (randomRIO)
import Control.Monad (foldM)
import Data.List (maximumBy)
import Data.Ord (comparing)

main :: IO ()
main = do
    print "Hello, Knockout Whist!"
    print $ pickTrumpSuit [Card Four Hearts, Card Two Clubs, Card Two Clubs, Card Six Diamonds]
    -- let trials = 5000000
    -- wins <- countWins trials
    -- putStrLn "Wins:"
    -- print wins
    -- putStrLn "Win Ratio:"
    -- print $ fromIntegral wins / fromIntegral trials


data Suit = Hearts | Diamonds | Clubs | Spades deriving (Eq, Enum, Bounded, Show)

data Rank = Two | Three | Four | Five | Six | Seven | Eight 
          | Nine | Ten | Jack | Queen | King | Ace
          deriving (Show, Eq, Ord, Enum, Bounded)

data Card = Card Rank Suit deriving (Eq, Show)
deck = [Card rank suit | rank <- [minBound .. maxBound], suit <- [minBound .. maxBound]]

pickNUniqueCards :: Int -> IO [Card]
pickNUniqueCards n = go n [] deck
  where
    go :: Int -> [Card] -> [Card] -> IO [Card]
    go 0 choices _ = return choices
    go n choices all = do
        i <- randomRIO (0, length all - 1)
        let choice = all !! i
        if choice `elem` choices
            then go n choices all --already chosen try again
            else go (n-1) (choice : choices) all

pickTrumpSuit :: [Card] -> Suit -- should give slightly more preference for more lower cards
pickTrumpSuit [] = Diamonds
pickTrumpSuit cards =
    let suits = [Hearts, Diamonds, Clubs, Spades]
        scores = map (\s -> sum [fromEnum r + 1 | Card r c <- cards, c == s]) suits
        argMax xs = fst $ maximumBy (comparing snd) (zip [0..] xs)
        i = argMax scores
    in suits !! i

playRound :: []
playTrick :: [[Card]] -> Suit -> Int -> Int
playTrick :: playersCards trump leader =
  let leaderChooseCard
    playerChooseCard



didTrumpWin :: [Card] -> Bool  --trump is the first in the list -- for the last round only
didTrumpWin (trumpCard : cards) = not $ any beatsTrump cards
    where 
        beatsTrump (Card r s) = (s == trumpSuit) && (r > trumpRank)
        Card trumpRank trumpSuit = trumpCard


countWins :: Int -> IO Int
countWins n = foldM go 0 [1..n]
  where
    go acc _ = do
      hand <- pickNUniqueCards 4-- [] deck
      return $ if didTrumpWin hand then acc + 1 else acc

